kzg_prover/benches/kzg.rs

```diff
@@ -0,0 +1,174 @@
+    let opening_grand_sum_bench_name = format!("<{}> opening grand sum", name);
+    let opening_user_bench_name = format!("<{}> opening user inclusion", name);
+    let verifying_grand_sum_bench_name = format!("<{}> verifying grand sum", name);
+    let verifying_user_bench_name = format!("<{}> verifying user inclusion", name);
+
+    let mut entries: Vec<Entry<N_CURRENCIES>> = vec![Entry::init_empty(); N_USERS];
+    let mut cryptos = vec![Cryptocurrency::init_empty(); N_CURRENCIES];
+    let _ =
+        parse_csv_to_entries::<&str, N_CURRENCIES>(csv_path, &mut entries, &mut cryptos).unwrap();
+
+    // Calculate total for all entry columns
+    let mut csv_total: Vec<BigUint> = vec![BigUint::from(0u32); N_CURRENCIES];
+    for entry in &entries {
+        for (i, balance) in entry.balances().iter().enumerate() {
+            csv_total[i] += balance;
+        }
+    }
+
+    c.bench_function(&range_check_bench_name, |b| {
+        b.iter_batched(
+            || circuit.clone(), // Setup function: clone the circuit for each iteration
+            |circuit| {
+                full_prover(&params, &pk, circuit, vec![vec![]]);
+            },
+            criterion::BatchSize::SmallInput, // Choose an appropriate batch size
+        );
+    });
+
+    let (zk_snark_proof, advice_polys, omega) = full_prover(&params, &pk, circuit, vec![vec![]]);
+
+    c.bench_function(&opening_grand_sum_bench_name, |b| {
+        b.iter_batched(
+            || 1..N_CURRENCIES + 1,
+            |balance_column_range| {
+                open_grand_sums::<N_CURRENCIES>(
+                    &advice_polys.advice_polys,
+                    &advice_polys.advice_blinds,
+                    &params,
+                    balance_column_range,
+                )
+            },
+            criterion::BatchSize::SmallInput,
+        );
+    });
+
+    // Generate a random user index
+    let get_random_user_index = || {
+        let user_range: std::ops::Range<usize> = 3..N_USERS;
+        OsRng.gen_range(user_range) as u16
+    };
+
+    c.bench_function(&opening_user_bench_name, |b| {
+        b.iter_batched(
+            || (get_random_user_index(), 0..N_CURRENCIES + 1),
+            |(user_index, column_range)| {
+                open_user_points::<N_CURRENCIES>(
+                    &advice_polys.advice_polys,
+                    &advice_polys.advice_blinds,
+                    &params,
+                    column_range,
+                    omega,
+                    user_index,
+                )
+            },
+            criterion::BatchSize::SmallInput,
+        );
+    });
+
+    // Open grand sum for benchmark verifying grand sum
+    let balance_column_range = 1..N_CURRENCIES + 1;
+    let grand_sums_batch_proof = open_grand_sums::<N_CURRENCIES>(
+        &advice_polys.advice_polys,
+        &advice_polys.advice_blinds,
+        &params,
+        balance_column_range.clone(),
+    );
+
+    c.bench_function(&verifying_grand_sum_bench_name, |b| {
```

Comment:
I think we need two separate benchmarks here: one for the grand sum opening and one for its verification. The reasoning is that the opening is performed by the Custodian and then made public so that the verification is performed by any interested party (e.g., Users).

Reply:
Has this been solved? @sifnoc @alxkzmn

Reply:
I forgot to reply for this.

I believe the benchmark you're referring to already exists in a separate section. You can find it in our repository here:
https://github.com/summa-dev/summa-solvency/blob/v2-bench/kzg_prover/benches/kzg.rs#L53-L57
```Rust
    c.bench_function(&opening_grand_sum_bench_name, |b| {
        b.iter_batched(
            || 1..N_CURRENCIES + 1,
            |balance_column_range| {
                open_grand_sums::<N_CURRENCIES>(
```
Please let me know if there's anything I might have missed